src/audio.c

```diff
@@ -0,0 +1,212 @@
+      return 0;
+   }
+   return devname;
+}
+
+static void printDevices(void) {
+   rsLog("Available devices:");
+   int max = SDL_GetNumAudioDevices(1);
+   for (int i = 0; i < max; ++i) {
+      rsLog(SDL_GetAudioDeviceName(i, 1));
+   }
+}
+
+static void prepareEncoder(RSAudio *audio, const RSConfig* config) {
+   RSAudioEncoder* audioenc = &audio->audioenc;
+   if (AACENC_OK != aacEncOpen(&(audioenc->aac_enc), 0, 0)) {
+      rsError("aacEncOpen failed");
+   }
+   aacEncoder_SetParam(audioenc->aac_enc, AACENC_TRANSMUX, 0);
+   aacEncoder_SetParam(audioenc->aac_enc, AACENC_AFTERBURNER, 1);
+   aacEncoder_SetParam(audioenc->aac_enc, AACENC_BITRATE, (UINT)config->audioBitrate);
+   aacEncoder_SetParam(audioenc->aac_enc, AACENC_SAMPLERATE, (UINT)config->audioSamplerate);
+   aacEncoder_SetParam(audioenc->aac_enc, AACENC_CHANNELMODE, (UINT)config->audioChannels);
+   if (AACENC_OK != aacEncEncode(audioenc->aac_enc, NULL, NULL, NULL, NULL)) {
+      rsError("aacEncEncode failed. Probably bad bitrate. Tried %d", config->audioBitrate);
+   }
+   aacEncInfo(audioenc->aac_enc, &(audio->audioenc.aac_info));
+
+   audioenc->data = 0;
+   audioenc->frame = 0;
+   audioenc->index = 0;
+   audioenc->size = audio->size;
+   audioenc->samplesPerFrame = (int)audioenc->aac_info.frameLength * config->audioChannels;
+   audioenc->frameSize = audioenc->samplesPerFrame * (int)sizeof(uint16_t);
+}
+
+static void prepareFrame(RSAudioEncoder *audioenc) {
+   int firstCopySize = (int)audioenc->frameSize;
+   int endData = (int)audioenc->index + audioenc->frameSize;
+   int diff = audioenc->size - endData;
+   if (diff >= 0) {
+      memcpy(audioenc->frame, audioenc->data + audioenc->index, (size_t)firstCopySize);
+      audioenc->index += firstCopySize;
+      return;
+   }
+   firstCopySize += diff;
+   memcpy(audioenc->frame, audioenc->data + audioenc->index, (size_t)firstCopySize);
+   memcpy(audioenc->frame + firstCopySize, audioenc->data, (size_t)-diff);
+   audioenc->index = -diff;
+}
+
+void rsAudioCreate(RSAudio *audio, const RSConfig *config) {
+   if (SDL_Init(SDL_INIT_AUDIO) < 0) {
+      rsError("SDL2 could not initialize! SDL Error: %s", SDL_GetError());
+   }
+   if (config->audioChannels != 1 && config->audioChannels != 2) {
+      rsLog("Only 1 or 2 channels are supported. Defaulting to 2");
+   }
+   if (pthread_spin_init(&sampleGetLock, PTHREAD_PROCESS_PRIVATE)) {
+      rsError("pthread_spin_init failed");
+   }
+   printDevices();
+   SDL_AudioDeviceID devId = 0;
+   SDL_AudioSpec ispec;
+   SDL_AudioSpec ospec;
+   SDL_zero(ispec);
+   SDL_zero(ospec);
+
+   ispec.freq = (int)config->audioSamplerate;
+   ispec.format = AUDIO_S16LSB;
+   ispec.channels = (Uint8)config->audioChannels;
+   ispec.samples = SAMPLES_PER_CALLBACK; //must be power of 2
+   ispec.callback = audioCallback;
+   ispec.userdata = audio;
+
+   const char* devname = getDeviceName(config);
+   devId = SDL_OpenAudioDevice(devname, 1, &ispec, &ospec, 0);
```

Comment:
Is there a way to get the default output device with your fork of SDL?

Reply:
You just call SDL_OpenAudioDevice with NULL devname to open the default device. getDeviceName returns 0  (will change to NULL) when the config string is "auto" or NULL itself

Reply:
Actually the strcmp was wrong, it would also return 0 on string being equal to auto, will change that

Reply:
I mean the default output device since I am assuming that will grab the default input device.

Reply:
It will grab whatever the default input device is, whether it's a microphone or an audio monitor is unknown. I could modify SDL to try get the first monitor it finds. That would require marking the device as such with some additional bool in the struct or something like that. I also doubt the official build of SDL2 will have a feature like that. They said they'll add extra argument to the getDevice functions that tells SDL2 if you want to filter out the audio monitors or not.
One other option is to query all available devices (something I already do) and look for "Monitor of" string, but as I said, I'm not sure if this string doesn't change between languages or who knows what other variables.